stdlib/camlinternalTLS.ml

```diff
@@ -0,0 +1,92 @@
+(*                 Stephen Dolan, University of Cambridge                 *)
+(*                   Tom Kelly, OCaml Labs Consultancy                    *)
+(*                                                                        *)
+(*   Copyright 2019 Indian Institute of Technology, Madras                *)
+(*   Copyright 2014 University of Cambridge                               *)
+(*   Copyright 2021 OCaml Labs Consultancy Ltd                            *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+type tls_state = Obj.t array
+
+let unique_value = Obj.repr (ref 0)
+
+external get_tls_state : unit -> tls_state = "%tls_get"
+external set_tls_state : tls_state -> unit = "caml_tls_set" [@@noalloc]
+
+type 'a t = int * (unit -> 'a)
+
+type key_initializer =
+  KI: 'a t * ('a -> 'a) -> key_initializer
+
+let key_counter = Atomic.make 0
+
+let parent_keys = Atomic.make ([] : key_initializer list)
+
+let rec add_parent_key ki =
+  let l = Atomic.get parent_keys in
+  if not (Atomic.compare_and_set parent_keys l (ki :: l))
+  then add_parent_key ki
+
+let make ?split_from_parent init_orphan : _ t =
+  let idx = Atomic.fetch_and_add key_counter 1 in
+  let k = (idx, init_orphan) in
+  begin match split_from_parent with
+  | None -> ()
+  | Some split -> add_parent_key (KI(k, split))
+  end;
+  k
+
+(* If necessary, grow the current domain's local state array such that [idx]
+ * is a valid index in the array. *)
+let maybe_grow idx =
+  let st = get_tls_state () in
+  let sz = Array.length st in
+  if idx < sz then st
+  else begin
+    let rec compute_new_size s =
+      if idx < s then s else compute_new_size (2 * s)
+    in
+    let new_sz = compute_new_size (max 8 sz) in
+    let new_st = Array.make new_sz unique_value in
+    Array.blit st 0 new_st 0 sz;
+    set_tls_state new_st;
+    new_st
+  end
+
+let set (idx, _init) x =
+  let st = maybe_grow idx in
+  (* [Sys.opaque_identity] ensures that flambda does not look at the type of
+   * [x], which may be a [float] and conclude that the [st] is a float array.
+   * We do not want OCaml's float array optimisation kicking in here. *)
+  st.(idx) <- Obj.repr (Sys.opaque_identity x)
+
+let get (idx, init) =
+  let st = maybe_grow idx in
+  let v = st.(idx) in
+  if v == unique_value then
+    let v' = Obj.repr (init ()) in
+    st.(idx) <- (Sys.opaque_identity v');
+    Obj.magic v'
+  else Obj.magic v
+
+let get_initial_keys () : (int * Obj.t) list =
+  List.map
+    (fun (KI ((idx, _) as k, split)) ->
+         (idx, Obj.repr (split (get k))))
+    (Atomic.get parent_keys)
```

Comment:
There is a reentrancy issue here if the calls to `get k` change the value of `parent_keys` by registering new keys that use inheritance.

I wrote the following not-too-horrible version that adds new keys in a loop.

```ocaml
let get_initial_keys () : (int * Obj.t) list =
  (* compute the key-value pair for the child from an initializer *)
  let kv_of_ki (KI ((idx, _) as k, split)) =
    (idx, Obj.repr (split (get k)))
  in
  (* We compute key-value pairs for all 'parent keys' in a loop,
     because the list of parent keys may grow during the
     computation. *)
  let rec loop acc last_parent_keys =
    let new_parent_keys = Atomic.get parent_keys in
    if new_parent_keys == last_parent_keys then acc
    else
      (* process all key initializers that are in [new_parent_keys]
         but not in [last_parent_keys], then loop again. *)
      let rec add_new_keys acc keys =
        if keys == last_parent_keys
        then loop acc new_parent_keys
        else match keys with
          | [] -> loop acc new_parent_keys
          | ki :: rest ->
            add_new_keys (kv_of_ki ki :: acc) rest
      in
      add_new_keys acc new_parent_keys
  in
  loop [] []
```

Reply:
Same as https://github.com/ocaml/ocaml/pull/12724#discussion_r1421708311 (?).